docs/proposals/20210720-kdm-template.md

```diff
@@ -0,0 +1,97 @@
+
+In this section, we will focus on the details of how to templating the control plane and would not discuss the details of how `NestedMachine` controller works. Assume that we have installed the `KubeadmControlPlaneController`, the `NestedMachineController`, and all necessary CAPI controllers on the meta cluster. We have also imported the `kubeadm` packages from the `kubernetes` module. Then, the templating process will include the following steps.
+
+1. The user applies the `KubeadmControlPlane` CR and all other necessary CAPI CRs.
+2. The `KubeadmControlPlane` controller keeps checking if the `NestedCluster` CR is ready.
+3. If yes, the `KubeadmControlPlane` controller use the `NestedMachineTemplate` to creates desired number of `NestedMachines` for the `NestedCluster` Infrastructure.
+4. The `KubeadmControlPlane` controller generates the `KubeadmConfig` CR based on the `KubeadmControlPlane` created in the first step. 
+5. The `KubeadmControlPlane` controller assigns the `KubeadmConfig` to the owned `NestedMachine` CR by setting the ObjectReference.
+6. The `NestedMachine` controller keeps checking if the `KubeadmConfig` has been assigned to the `NestedMachine` CR.
+7. If yes, the `NestedMachine` controller calls functions from the `kubeadm` package to generate the static pod template.
+8. The `NestedMachine` controller extracts the `PodTemplateSpec` from the static pod template generated in the sixth step and creates the StatefulSet template.
+
+![Templating the control plane](images/nestedcontrolplane/kdm-template.svg)
+
+### User Stories
+
+1. As the `NestedMachine` controller, I would like to create a `NestedControlPlane` with the same templates as the kubernetes control plane created by the `kubeadm`.
+2. As a user, I would like to customize the command line arguments of my `NestedControlPlane`'s components through a `KubeadmControlPlane` CR.
+3. As a user, I would like to customize the pod volumes of my `NestedControlPlane`'s components through a `KubeadmControlPlane` CR.
+
+### Implementation Details
+
+The provisioning of the control plane requires the involvement of the `NestedMachine` controller. We will discuss the `NestedMachine` controller and the complete provisioning process in another CAEP. 
+This CAEP will focus on how to template the control plane in a more "CAPI" way, i.e., using `KubeadmControlPlane` provider and `kubeadm` libs from the `kubernetes` module.  
+
+- Use the existing `NestedControlPlane` implemenation for testing.
+- Create `KubeadmConfig` CR manually.
+- Implement the generic functionality of extracting the`kubeadmapi.ClusterConfig` from the `KubeadmConfig` CR.
+- Copy the templating functions from the `kubeadm` libs in the `Kubernetes` module (we will import the functions instead of copying, once `kubeadm` is independent from the `Kubernetes`).
+- Import the `kubeadm` lib from the `Kubernetes` module and call related functions to get the static pod template for each component.
```

Comment:
Did we want to call out that we're going to copy the code out of tree instead of import based on @vincepri's feedback?